gcloud-java-bigquery/src/main/java/com/google/gcloud/bigquery/JobConfiguration.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.bigquery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.MoreObjects.ToStringHelper;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+/**
+ * Base class for a BigQuery job configuration.
+ */
+public abstract class JobConfiguration implements Serializable {
+
+  private static final long serialVersionUID = -548132177415406526L;
+
+  private final Type type;
+
+  /**
+   * Type of a BigQuery Job.
+   */
+  enum Type {
+    /**
+     * A Copy Job copies an existing table to another new or existing table. Instances of
+     * {@code JobConfiguration} for this type are implemented by {@link CopyJobConfiguration}.
+     */
+    COPY,
+    /**
+     * An Extract Job exports a BigQuery table to Google Cloud Storage. Instances of
+     * {@code JobConfiguration} for this type are implemented by {@link ExtractJobConfiguration}.
+     */
+    EXTRACT,
+    /**
+     * A Load Job loads data from one of several formats into a table.  Instances of
+     * {@code JobConfiguration} for this type are implemented by {@link LoadJobConfiguration}.
+     */
+    LOAD,
+    /**
+     * A Query Job runs a query against BigQuery data.  Instances of
+     * {@code JobConfiguration} for this type are implemented by {@link QueryJobConfiguration}.
+     */
+    QUERY
+  }
+
+  /**
+   * Base builder for job configurations.
+   *
+   * @param <T> the job configuration type
+   * @param <B> the job configuration builder
+   */
+  public abstract static class Builder<T extends JobConfiguration, B extends Builder<T, B>> {
+
+    private Type type;
+
+    protected Builder(Type type) {
+      this.type = checkNotNull(type);
+    }
+
+    @SuppressWarnings("unchecked")
+    protected B self() {
+      return (B) this;
+    }
+
+    B type(Type type) {
+      this.type = checkNotNull(type);
+      return self();
+    }
+
+    /**
+     * Creates an object.
+     */
+    public abstract T build();
+  }
+
+  protected JobConfiguration(Builder builder) {
```

Comment:
if we make this package scope, we make sure that we will never need to deal with a configuration that we didn't expect.